crates/sui-storage/src/write_ahead_log.rs

```diff
@@ -0,0 +1,292 @@
+    /// list if not explicitly committed or aborted.
+    ///
+    /// The caller is responsible for running each tx to completion.
+    ///
+    /// Recoverable TXes will remain in the on-disk log until they are explicitly committed or
+    /// aborted.
+    #[must_use]
+    fn take_recoverable_txes(&'a self) -> Vec<Self::Guard>;
+
+    /// Get the data associated with a given digest - returns an error if no such transaction is
+    /// currently open.
+    /// Requires a TxGuard to prevent asking about transactions that aren't in the log.
+    fn get_tx_data(&self, g: &Self::Guard) -> SuiResult<C>;
+}
+
+pub struct DBTxGuard<'a, C: Serialize + DeserializeOwned> {
+    tx: TransactionDigest,
+    wal: &'a DBWriteAheadLog<C>,
+    dead: bool,
+}
+
+impl<'a, C> DBTxGuard<'a, C>
+where
+    C: Serialize + DeserializeOwned,
+{
+    fn new(tx: &TransactionDigest, wal: &'a DBWriteAheadLog<C>) -> Self {
+        Self {
+            tx: *tx,
+            wal,
+            dead: false,
+        }
+    }
+}
+
+impl<'a, C> TxGuard<'a> for DBTxGuard<'a, C>
+where
+    C: Serialize + DeserializeOwned,
+{
+    fn tx_id(&self) -> TransactionDigest {
+        self.tx
+    }
+
+    fn commit_tx(mut self) -> SuiResult {
+        self.dead = true;
+        self.wal.commit_tx(&self.tx)
+    }
+
+    fn abort_tx(mut self) -> SuiResult {
+        self.dead = true;
+        self.wal.abort_tx(&self.tx)
+    }
+}
+
+impl<C> Drop for DBTxGuard<'_, C>
+where
+    C: Serialize + DeserializeOwned,
+{
+    fn drop(&mut self) {
+        if !self.dead {
+            let tx = self.tx;
+            error!(digest = ?tx, "DBTxGuard dropped without explicit commit/abort");
+            self.wal.implicit_drop_tx(&tx);
+        }
+    }
+}
+
+// A WriteAheadLog implementation built on rocksdb.
+pub struct DBWriteAheadLog<C> {
+    log: DBMap<TransactionDigest, C>,
+
+    // Can't use tokio Mutex - must be accessible synchronously from drop trait.
+    recoverable_txes: Mutex<Vec<TransactionDigest>>,
+
+    // Guards the get/set in begin_tx
+    mutex_table: MutexTable<TransactionDigest>,
+}
+
+const MUTEX_TABLE_SIZE: usize = 1024;
+
+impl<C> DBWriteAheadLog<C>
+where
+    C: Serialize + DeserializeOwned,
+{
+    pub fn new<P: AsRef<Path>>(path: P) -> Self {
+        let (options, _) = default_db_options(None);
+        let db = {
+            let path = &path;
+            let db_options = Some(options.clone());
+            let opt_cfs: &[(&str, &rocksdb::Options)] = &[("tx_write_ahead_log", &options)];
+            typed_store::rocks::open_cf_opts(path, db_options, opt_cfs)
+        }
+        .expect("Cannot open DB.");
+
+        let log: DBMap<TransactionDigest, C> =
+            DBMap::reopen(&db, Some("tx_write_ahead_log")).expect("Cannot open CF.");
+
+        // Read in any digests that were left in the log, e.g. due to a crash.
+        let recoverable_txes: Vec<_> = log.iter().map(|(tx, _)| tx).collect();
```

Comment:
Make a note that this is guaranteed to be a small list since we delete entries after we commit / abort.

Reply:
Well, its not guaranteed - if we suddenly start erroring out of every tx write this will grow without bound. You can imagine this happening once we are writing to remote stores instead of doing everything in process, if one of the remote stores goes down.

However, if this happens, the validator is essentially broken anyway and should probably stop accepting requests. I'm not sure this is the right place to worry about this problem.

Reply:
We could add a TODO to add metrics here to monitor if that situation happens.  Of course, we should have plenty of other monitoring activate in that case.

@mystenmark could you add to the module docs here that the state is intended to be small, just for in progress transactions?